src/DotLiquid/Context.cs

```diff
@@ -422,29 +433,42 @@ private object Resolve(string key, bool notifyNotFound = true)
         /// the hierarchy
         /// </summary>
         /// <param name="key"></param>
+        /// <param name="variable"></param>
         /// <returns></returns>
-        private object FindVariable(string key)
+        private bool TryFindVariable(string key, out object variable)
         {
+            bool foundVariable = false;
             Hash scope = Scopes.FirstOrDefault(s => s.ContainsKey(key));
-            object variable = null;
+            object ret = null;
             if (scope == null)
             {
                 foreach (Hash e in Environments)
-                    if ((variable = LookupAndEvaluate(e, key)) != null)
+                {
+                    foundVariable = TryEvaluateHashOrArrayLikeObject(e, key, out ret);
+                    if (foundVariable)
                     {
                         scope = e;
                         break;
                     }
+                }
+
+                if (scope == null)
+                {
+                    scope = Environments.LastOrDefault() ?? Scopes.Last();
+                    foundVariable = TryEvaluateHashOrArrayLikeObject(scope, key, out ret);
+                }
+            }
+            else
+            {
+                foundVariable = TryEvaluateHashOrArrayLikeObject(scope, key, out ret);
             }
-            scope = scope ?? Environments.LastOrDefault() ?? Scopes.Last();
-            variable = variable ?? LookupAndEvaluate(scope, key);
 
-            variable = Liquidize(variable);
+            variable = Liquidize(ret);

```

Comment:
I think the expected contract from a TryXyz method is that the return object is assigned null if the attempt fails. Here it assigns even if `foundVariable` is false.

Reply:
Probably doesn't hurt because ret is null in that case. But makes sense.

Reply:
Actually, you can't skip assignment - compiler throws an error - The out parameter 'variable' must be assigned to before control leaves the current method

Reply:
Yes which is why engineers typically assign the out variable to `null` for when there isn't a match.

Reply:
Which is exactly what will happen, the underlying variable will be null, liquidized and still be null. I don't think there is a case where foundVariable is false and the variable is not null.

Reply:
oh!